xarray/backends/zarr.py

```diff
@@ -0,0 +1,474 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+import functools
+import warnings
+from itertools import product
+from collections import MutableMapping
+import operator
+
+from .. import Variable
+from ..core import indexing
+from ..core.utils import (FrozenOrderedDict, close_on_error, HiddenKeyDict,
+                          NdimSizeLenMixin,DunderArrayMixin)
+from ..core.pycompat import (iteritems, bytes_type, unicode_type, OrderedDict,
+                             basestring)
+
+from .common import (WritableCFDataStore, AbstractWritableDataStore,
+                     DataStorePickleMixin)
+
+from .. import conventions
+
+# most of the other stores have some kind of wrapper class like
+# class BaseNetCDF4Array(NdimSizeLenMixin, DunderArrayMixin):
+# class H5NetCDFArrayWrapper(BaseNetCDF4Array):
+# class NioArrayWrapper(NdimSizeLenMixin, DunderArrayMixin):
+# we problaby need something like this
+
+# the first question is whether it should be based on BaseNetCDF4Array or
+# NdimSizeLenMixing?
+
+# or maybe we don't need wrappers at all? probably not true
+
+class ZarrArrayWrapper(NdimSizeLenMixin, DunderArrayMixin):
+    def __init__(self, variable_name, datastore):
+        self.datastore = datastore
+        self.variable_name = variable_name
+
+        array = self.get_array()
+        self.shape = array.shape
+
+        dtype = array.dtype
+        if dtype is str:
+            # use object dtype because that's the only way in numpy to
+            # represent variable length strings; it also prevents automatic
+            # string concatenation via conventions.decode_cf_variable
+            dtype = np.dtype('O')
```

Comment:
I basically just copied this from `BaseNetCDF4Array` and I don't really know if it makes sense here.